File: 121.BestTimeToBuyAndSellStock/MarkCSIE/BestTimeToBuyAndSellStock/src/BestTimeToBuyAndSellStock.java

public class BestTimeToBuyAndSellStock {

    public static void main(String[] args) {
        BestTimeToBuyAndSellStock test = new BestTimeToBuyAndSellStock();
        System.out.println(test.maxProfit(new int[]{2, 1, 3, 4, 5, 3, 1, 8}));
    }

    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int length = prices.length;
        int[] minIndex = new int[length];
        if (length != 0) {
            minIndex[0] = prices[0];
            for (int i = 1; i < length; i++) {
                minIndex[i] = Math.min(minIndex[i - 1], prices[i]);
                maxProfit = Math.max(maxProfit, prices[i] - minIndex[i]);
            }
        }
        return maxProfit;
    }
}

File: 121.BestTimeToBuyAndSellStock/Ddfish/BestTimeToBuyAndSellStock.swift
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
    
       if(prices.count == 0){
            return 0;
        }
   
        var min = prices[0];
        var profit = 0;
     
        for price in prices{
            if(min > price){
                min = price;
            }else if(price - min > profit){
                profit = price - min;
            }
        
       } 
        
        return profit;
        
    }
}
